### ISSUE TITLE
Custom Ordinal Multi-Logloss

### ISSUE THREAD

**Issue by akaniklaus** [Microsoft/LightGBM#1742]
Dear Community, I want to have a custom loss function where the calculated multi_logloss is weighted with np.square(y_pred -  y_test) + 1.0. I checked the advaced example but couldn't succeed doing it. I would very much welcome if someone can help. Thank you in advance for your time and consideration.

**Comment by akaniklaus** [comment]
Is there any possible way to handle regression of imbalanced data in LightGBM? As my regression labels are discrete, I would like to implement a custom loss function that acts as a weighted mean square error.

**Comment by guolinke** [comment]
I think you are trying to solve the multi-class problem by regression.
If so, you should set the `num_class=1`.

As for the objective, it is something like:
```python
def object(preds, train_data):
    labels = train_data.get_label()
    weights = np.take(class_weights, labels.astype(int))
    grad = (preds - labels)**2 * weights 
    hess = np.ones(grad.shape)
    return grad.flatten('F'), hess.flatten('F')
```

And you may need to write a customed function to convert the regression prediction to class prediction by yourself.

**Comment by akaniklaus** [comment]
@guolinke Hello, thank you very much for your response! I have tried this but the regression returns wrong values (too much different from when I use the default loss function). The dataset that I am currently employing is not that unbalanced so I guess there is a mistake at grad or hess calculation. Note: I have also tried to calculate class weights manually and feed them via np.take as you have done.

```
from sklearn.utils.class_weight import compute_sample_weight

def fobj(preds, train_data):
    labels = train_data.get_label().astype(int)
    weights = compute_sample_weight('balanced', labels.astype(int))
    grad = (preds - labels)**2
    grad = grad * weights
    return grad, weights
```

**Comment by akaniklaus** [comment]
@guolinke Ok, I believe that I found out the problem. I have multiplied the "grad" by np.sign(preds-labels) after figuring out that grad also needs the direction (sign). Thank you very much again. Have a nice day.
